encoding.go

```diff
@@ -71,74 +100,76 @@ func ParseStatelessNode(serialized []byte, depth byte, comm []byte) (VerkleNode,
+		if bit(bitlist, i) {
+			if offset+LeafValueSize > len(serialized) {
+				return nil, fmt.Errorf("verkle payload is too short, need at least %d and only have %d, payload = %x (%w)", offset+32, len(serialized), serialized, errSerializedPayloadTooShort)
 			}
-			values[i] = serialized[offset : offset+32]
-			offset += 32
+			values[i] = serialized[offset : offset+LeafValueSize]
+			offset += LeafValueSize
 		}
 	}
-	if NodeWidth != len(values) {
-		return nil, fmt.Errorf("invalid number of nodes in decoded child expected %d, got %d", NodeWidth, len(values))
-	}
-	ln := NewLeafNodeWithNoComms(serialized[1:32], values[:])
+	ln := NewLeafNodeWithNoComms(serialized[leafSteamOffset:leafSteamOffset+StemSize], values[:])
 	ln.setDepth(depth)
 	ln.c1 = new(Point)
-	ln.c1.SetBytesTrusted(serialized[64:96])
+	ln.c1.SetBytesUncompressed(serialized[leafC1CommitmentOffset:leafC1CommitmentOffset+SerializedPointSize], true)
 	ln.c2 = new(Point)
-	ln.c2.SetBytesTrusted(serialized[96:128])
+	ln.c2.SetBytesUncompressed(serialized[leafC2CommitmentOffset:leafC2CommitmentOffset+SerializedPointSize], true)
 	ln.commitment = new(Point)
-	ln.commitment.SetBytesTrusted(comm)
+	ln.commitment.SetBytes(comm, true)
 	return ln, nil
 }
 
-func deserializeIntoStateless(bitlist []byte, raw []byte, depth byte, comm []byte) (*StatelessNode, error) {
+func deserializeIntoStateless(bitlist []byte, raw []byte, depth byte, comm SerializedPointCompressed) (*StatelessNode, error) {
 	// GetTreeConfig caches computation result, hence
 	// this op has low overhead
 	n := NewStateless()
 	n.setDepth(depth)
 	indices := indicesFromBitlist(bitlist)
-	if len(raw)/32 != len(indices) {
+	if len(raw)/SerializedPointSize != len(indices) {
 		return nil, ErrInvalidNodeEncoding
 	}
 	for i, index := range indices {
-		n.unresolved[byte(index)] = raw[i*32 : (i+1)*32]
+		n.unresolved[byte(index)] = raw[i*SerializedPointSize : (i+1)*SerializedPointSize]
 	}
 	n.commitment = new(Point)
-	n.commitment.SetBytesTrusted(comm)
+	n.commitment.SetBytes(comm, true)
 	return n, nil
 }
 
-func CreateInternalNode(bitlist []byte, raw []byte, depth byte, comm []byte) (*InternalNode, error) {
+func CreateInternalNode(bitlist []byte, raw []byte, depth byte, comm SerializedPointCompressed) (*InternalNode, error) {
 	// GetTreeConfig caches computation result, hence
 	// this op has low overhead
 	n := (newInternalNode(depth)).(*InternalNode)
 	indices := indicesFromBitlist(bitlist)
-	if len(raw)/32 != len(indices) {
+
+	if len(raw)/SerializedPointSize != len(indices) {
 		return nil, ErrInvalidNodeEncoding
 	}
+
+	freelist := make([]HashedNode, len(indices))
 	for i, index := range indices {
-		n.children[index] = &HashedNode{commitment: raw[i*32 : (i+1)*32]}
+		freelist[i].serialized = raw[i*SerializedPointSize : (i+1)*SerializedPointSize]
+		n.children[index] = &freelist[i]
 	}
```

Comment:
This innocent change saved billions of allocations (I mean, _literally_), which greatly impacted the replay benchmark (i.e: double digit % improvement).
What happened here is that each `&HashedNode{..}` call would make a heap allocation, since that's a pointer stored as a `VerkleNode` node.

That's unfortunate, since we're creating _a lot_ of `HashedNodes`, and that means a lot of `mallocgc` calls plus a lot of GC pressure.

So, what I did is simple. We already know here how many `HashedNode`s we'll need. I make a single slice allocation of that size, and then initialize from there. That means only 1 alloc vs worst-case 256. Regarding "freeing" semantics, all is fine, since all _HashNodes_ are children of the same parent, so all of them will be "dropped" at the same time liberating the slice memory.

All this is is another example of why it's useful to see pprofs of e2e benchmarks; not the usual thing you'll imagine it will save a billion allocations and save so many seconds of CPU.